
# 20260724 Adriaan van Os <adriaan@adriaan.biz>
# patch for fpc-3.3.1 (against commit
# https://github.com/fpc/FPCSource/commit/9e4ac8d7ea2a87dd909d829122d0b3b66f42ed02)
# to implement asynchronous assembly, to speed up fpc compilation


diff -ur a/compiler/assemble.pas b/compiler/assemble.pas
--- a/compiler/assemble.pas	2026-07-06 22:25:13
+++ b/compiler/assemble.pas	2026-07-24 08:52:40
@@ -257,6 +257,9 @@
 {$ifdef hasunix}
       unix,
 {$endif}
+{$ifdef async_assembly}
+      baseunix,unixutil,
+{$endif}
       cutils,cfileutl,
 {$ifdef memdebug}
       cclasses,
@@ -364,7 +367,8 @@
       var
         g : file;
       begin
-        if cs_asm_leave in current_settings.globalswitches then
+        if ( cs_asm_leave in current_settings.globalswitches)
+        or ( cs_asm_async in current_settings.globalswitches) then
          exit;
         if cs_asm_extern in current_settings.globalswitches then
          AsmRes.AddDeleteCommand(owner.AsmFileName)
@@ -682,13 +686,22 @@
       var
         f : file;
         FileAge : longint;
+        theChildPID : cint;
       begin
         AsmFlush;
 {$ifdef hasunix}
         if owner.DoPipe then
           begin
-            if PClose(outfile) <> 0 then
-              GenerateError;
+            if cs_asm_async in current_settings.globalswitches then
+              begin
+                PCloseNoWait(outfile, theChildPID);
+                RegisterAsyncAssembly( theChildPID, '')
+              end
+            else
+              begin
+                if PClose(outfile) <> 0 then
+                  GenerateError
+              end
           end
         else
 {$endif}
@@ -926,8 +939,55 @@
         end;
       end;
 
+ function AssembleOutOfProcess
+        (     theProgramPath  : AnsiString;
+              theArguments  : RawByteString;
+          var theChildPID : cint;
+          var theAsmFile  : AnsiString): cint;
+      var
+        theErr  : cint;
+        thePid  : cint;
+        thePPArguments  : PPAnsiChar;
+        thePPChar : PPAnsiChar;
+    begin
+      theErr  := 0;
+      theChildPID := 0;
+      theAsmFile  := '';
+      UniqueString
+        ( theArguments);
+      thePPArguments  := StringToPPChar( theArguments, 1);
+      thePPArguments^ := @theProgramPath[ 1];
+      thePPChar := thePPArguments;
+      while thePPChar^ <> nil do
+      begin
+        theAsmFile  := thePPChar^;
+        thePPChar := PPAnsiChar( UInt64( thePPChar) + SizeOf( Pointer))
+      end;
+      thePID  := fpFork;
+      if thePID = 0
+        then begin
+          { we are in the child }
+          theErr  := FpExecV( theProgramPath, thePPArguments);
+          Halt
+            ( 127) { as FpExecV should not return }
+        end
+        else
+      if thePID = - 1
+        then theErr := fpGetErrno
+        else begin
+         { we are in the parent }
+          theChildPID := thePID;
+        end;
+      AssembleOutOfProcess  := theErr
+    end;
 
+
     Function TExternalAssembler.DoAssemble:boolean;
+{$ifdef async_assembly}
+      var
+        theChildPID : cint;
+        theAsmFile  : AnsiString;
+{$endif}
       begin
         result:=true;
         if DoPipe then
@@ -944,7 +1004,19 @@
          end;
 
         repeat
-          result:=CallAssembler(FindAssembler,MakeCmdLine)
+        {$ifdef async_assembly}
+          if cs_asm_async in current_settings.globalswitches then
+            begin
+              result := AssembleOutOfProcess( FindAssembler, MakeCmdLine, theChildPID, theAsmFile) = 0;
+              if result then
+                if cs_asm_leave in current_settings.globalswitches then
+                  RegisterAsyncAssembly( theChildPID, '')
+                else
+                  RegisterAsyncAssembly( theChildPID, theAsmFile)
+            end
+          else
+        {$endif}
+            result:=CallAssembler(FindAssembler,MakeCmdLine)
         until not(result) or not RerunAssembler;
         if result then
           writer.RemoveAsm

diff -ur a/compiler/fpcdefs.inc b/compiler/fpcdefs.inc
--- a/compiler/fpcdefs.inc	2026-07-06 22:25:13
+++ b/compiler/fpcdefs.inc	2026-07-21 14:21:59
@@ -464,3 +464,11 @@
   {$define SUPPORT_GET_FRAME}
 {$endif}
 
+{$ifdef darwin}
+{$define async_assembly}
+{$endif}
+
+{$define true}
+{$undef false}
+
+
diff -ur a/compiler/globals.pas b/compiler/globals.pas
--- a/compiler/globals.pas	2026-07-06 22:25:13
+++ b/compiler/globals.pas	2026-07-24 08:55:32
@@ -769,7 +769,36 @@
   var
     AllowedFilenameTransFormations : tfilenametransformations = AllTransformations;
 
+{$ifdef async_assembly}
+  const
+    kAsyncAsmBufferSize = 8;
+  type
+    TAsyncAsmPath = AnsiString;
+    PAsyncAsmData = ^TAsyncAsmData;
+    TAsyncAsmData =
+      record { circular buffer }
+        cbNum : Int32;
+          { round buffer number of entries (between 0 and kAsyncAsmBufferSize) }
+        cbNextIndex : Int32;
+          { round buffer index of next PID entry}
+        cbEntries : array[ 1..kAsyncAsmBufferSize] of
+          record
+            ePid  : TPID;
+              { PID of the asynchronous external assembly process, e.g. clang }
+            eAsmFile  : TAsyncAsmPath;
+              { path of the assembly file, if it needs to be removed }
+          end
+      end;
 
+  var
+    gAsyncAsmData : TAsyncAsmData;
+
+  procedure InitAsyncAssembly;
+  procedure DoneAsyncAssembly;
+  procedure RegisterAsyncAssembly( theChildPID: pid_t; theAsmFileToRemove: AnsiString);
+  procedure WaitForAsyncAssembly; { to be called before linking starts }
+{$endif}
+
 implementation
 
 {$if defined(macos)}
@@ -1712,6 +1741,9 @@
    procedure DoneGlobals;
      begin
        calldoneprocs;
+     {$ifdef async_assembly}
+       DoneAsyncAssembly;
+     {$endif}
        librarysearchpath.Free;
        librarysearchpath := nil;
        unitsearchpath.Free;
@@ -1820,8 +1852,99 @@
         { enable all features by default }
         features:=[low(Tfeature)..high(Tfeature)];
 
+      {$ifdef async_assembly}
+        InitAsyncAssembly;
+      {$endif}
+
         callinitprocs;
      end;
+
+{$ifdef async_assembly}
+procedure InitAsyncAssembly;
+begin
+  with gAsyncAsmData do
+  begin
+    cbNum := 0;
+    cbNextIndex := 1;
+    FillChar
+      ( cbEntries, SizeOf( cbEntries), 0)
+  end
+end;
+{$endif}
+
+{$ifdef async_assembly}
+procedure DoneAsyncAssembly;
+begin
+end;
+{$endif}
+
+{$ifdef async_assembly}
+ function waitpid
+        (     thePid  : pid_t;
+          var theStatLoc  : cint;
+              theOptions  : cint): pid_t; cdecl; external;
+
+ function unlink
+        (     thePath : PChar): cint; cdecl; external;
+
+procedure WaitForAsyncAssemblyEntry
+        (       theIndex  : Int32);
+var
+  theResult : pid_t;
+  theStat : cint;
+begin
+  with gAsyncAsmData.cbEntries[ theIndex] do
+  begin
+    if ePid <> 0
+      then begin
+        theResult := waitpid( ePid, theStat, 0);
+        ePid := 0
+      end;
+    if Length( eAsmFile) > 0
+      then begin
+        theResult := unlink
+          ( @eAsmFile)
+      end
+  end
+end;
+
+procedure WaitForAsyncAssembly;
+var
+  theIndex  : Int32;
+begin
+  with gAsyncAsmData do
+  begin
+    for theIndex  := 1 to cbNum do
+      WaitForAsyncAssemblyEntry
+        ( theIndex);
+    cbNum := 0;
+    cbNextIndex := 1
+  end
+end;
+{$endif}
+
+{$ifdef async_assembly}
+procedure RegisterAsyncAssembly
+        (     theChildPID : pid_t;
+              theAsmFileToRemove  : AnsiString);
+    begin
+      with gAsyncAsmData do
+      begin
+        if cbNum = kAsyncAsmBufferSize
+          then WaitForAsyncAssemblyEntry
+             ( cbNextIndex)
+          else cbNum  := cbNum + 1;
+        with cbEntries[ cbNextIndex] do
+        begin
+          ePid  := theChildPid;
+          eAsmFile  := theAsmFileToRemove
+        end;
+        if cbNextIndex = kAsyncAsmBufferSize
+          then cbNextIndex  := 1
+          else cbNextIndex  := cbNextIndex + 1
+      end
+    end;
+{$endif}
 
 initialization
 {$ifdef LLVM}

diff -ur a/compiler/globtype.pas b/compiler/globtype.pas
--- a/compiler/globtype.pas	2026-07-08 11:51:30
+++ b/compiler/globtype.pas	2026-07-19 18:40:03
@@ -229,6 +229,7 @@
          { assembling }
          cs_asm_leave,cs_asm_extern,cs_asm_pipe,cs_asm_source,cs_asm_rtti_source,
          cs_asm_regalloc,cs_asm_tempalloc,cs_asm_nodes,cs_asm_pre_binutils_2_25,
+         cs_asm_async,
          { linking }
          cs_link_nolink,cs_link_static,cs_link_smart,cs_link_shared,cs_link_deffile,
          cs_link_strip,cs_link_staticflag,cs_link_on_target,cs_link_extern,cs_link_opt_vtable,

diff -ur a/compiler/options.pas b/compiler/options.pas
--- a/compiler/options.pas	2026-07-08 11:51:30
+++ b/compiler/options.pas	2026-07-19 18:41:22
@@ -2449,12 +2449,17 @@
 procedure TOption.Interpret_A_U(opt, more: TCmdStr);
 
 begin
-  if CompareText(More,'DEFAULT') = 0 then
-    paratargetasm:=as_default
+  if more = 'async' then
+    include(init_settings.globalswitches,cs_asm_async)
   else
-    paratargetasm:=find_asm_by_string(More);
-  if paratargetasm=as_none then
-    IllegalPara(opt);
+    begin
+      if CompareText(More,'DEFAULT') = 0 then
+        paratargetasm:=as_default
+      else
+        paratargetasm:=find_asm_by_string(More);
+      if paratargetasm=as_none then
+        IllegalPara(opt);
+    end;
 end;
 
 

diff -ur a/compiler/pmodules.pas b/compiler/pmodules.pas
--- a/compiler/pmodules.pas	2026-07-06 22:25:13
+++ b/compiler/pmodules.pas	2026-07-20 01:04:01
@@ -2322,6 +2322,11 @@
                begin
                  { create global resource file by collecting all resource files }
                  CollectResourceFiles;
+
+              {$ifdef async_assembly}
+                WaitForAsyncAssembly;
+              {$endif}
+
                  { write .def file }
                  if (cs_link_deffile in current_settings.globalswitches) then
                    deffile.writefile;
@@ -2378,6 +2383,9 @@
       begin
             { create global resource file by collecting all resource files }
             CollectResourceFiles;
+          {$ifdef async_assembly}
+            WaitForAsyncAssembly;
+          {$endif}
             { write .def file }
             if (cs_link_deffile in current_settings.globalswitches) then
              deffile.writefile;


diff -ur a/compiler/utils/ppuutils/ppudump.pp b/compiler/utils/ppuutils/ppudump.pp
--- a/compiler/utils/ppuutils/ppudump.pp	2026-07-08 11:51:30
+++ b/compiler/utils/ppuutils/ppudump.pp	2026-07-19 18:39:43
@@ -2375,6 +2375,7 @@
         'Add temporary  allocation into assmebler files', {cs_asm_tempalloc}
         'Add node information into assembler files', {cs_asm_nodes}
         'Adapt assembler call to GNU version <= 2.25', {cs_asm_pre_binutils_2_25}
+        'Assemble asynchronously, out-of-process (AvO)', {cs_asm_async}
          { linking }
         'Skip linking stage', {cs_link_nolink}
         'Link static', {cs_link_static}

diff -ur a/rtl/darwin/unxfunc.inc b/rtl/darwin/unxfunc.inc
--- a/rtl/darwin/unxfunc.inc	2026-07-06 22:25:13
+++ b/rtl/darwin/unxfunc.inc	2026-07-21 15:21:03
@@ -26,6 +25,18 @@
 { closed our side, Now wait for the other - this appears to be needed ?? }
   pl:=@(filerec(f).userdata[2]);
   pclose := WaitProcess(pl^);
+end;
+
+Procedure PCloseNoWait(Var F:file; Var thePID: cint);
+var
+  pl : ^cint;
+  res: cint;
+begin
+  repeat
+    res:=fpclose(filerec(F).Handle);
+  until (res<>-1) or (fpgeterrno<>ESysEINTR);
+  pl:=@(filerec(f).userdata[2]);
+  thePID := pl^
 end;
 
 Function PClose(Var F:text) :cint;

diff -ur a/rtl/unix/unxovlh.inc b/rtl/unix/unxovlh.inc
--- a/rtl/unix/unxovlh.inc	2026-07-06 22:25:13
+++ b/rtl/unix/unxovlh.inc	2026-07-21 15:21:19
@@ -15,5 +15,6 @@
 
 Function PClose(Var F:file) : cint;
 Function PClose(Var F:text) : cint;
+Procedure PCloseNoWait(Var F:file; Var thePID: cint);
 Function fpStatFS(Path:ansistring;Info:PStatfs):cint;
 
